Submit event excuses

MichaelJamesParsons opened the pull request:
Ready for review. I left out success/error indicators for now to keep the PR small(er). I'll add a snackbar to both the excuses and file entry UI's soon.

    ```
    git rebase -i $(git merge-base HEAD master)
    git push origin NAME_OF_BRANCH --force-with-lease
    ```

![excuses_1](https://user-images.githubusercontent.com/2666285/52394427-1d19ca80-2a5f-11e9-837b-26b071a70e73.png)
![excuses_2](https://user-images.githubusercontent.com/2666285/52394428-1d19ca80-2a5f-11e9-97b2-2d2274afee1c.png)

## Migration Notes
Static12 allows null value for the event date, date created, and date modified. These fields are required in titan. To avoid complications while migrating excuses, we must either remove invalid records or set a default date.

ELD commented on titan-server/migrations/2019-02-03-013951_create_excuses_table/up.sql:
```diff
@@ -0,0 +1,12 @@
+CREATE TABLE user_event_excuses (
```
The fields that aren't nullable should be marked as such in the migration. This prevents Diesel from generating schemas that are incompatible with the type system. This is my view of nullable versus not nullable types in this database table:

|Column|Type|Nullable?|
|--|--|--|
|static12_id|INT|No|
|event_type|VARCHAR(255)|No|
|event_date|DATETIME|No|
|comments|TEXT|No|
|ack_user_id|INT|Yes|
|ack_date|DATETIME|Yes|
|ack_comments|TEXT|Yes|
|date_created|DATETIME|No|
|date_modified|DATETIME|No|

ELD commented on titan-server/src/accounts/excuses.rs:
```diff
@@ -0,0 +1,45 @@
+use crate::models;
+use crate::schema;
+
+pub const EVENT_TYPES: &'static [&'static str] = &[
```
Should this be moved to a separate table that can have event types expanded upon (or shrunk down) depending on the changes to event types in the clan?

This seems logically inconsistent with how we treat file entry types as a joinable table, but this is being treated as a static listing.

MichaelJamesParsons replied:
That makes sense. I'll go ahead and add an `event_type` entity.

ELD commented on titan-server/src/schema.rs:
```diff
@@ -222,6 +222,21 @@ table! {
 }
 
+table! {
+    user_event_excuses (id) {
```
Refer to the above comment in the SQL migration, but anything that isn't wrapped in `Nullable<T>` in diesel's schema should be marked `NOT NULL` in the actual SQL.

ELD commented on titan-server/src/schema.rs:
```diff
@@ -222,6 +222,21 @@ table! {
+table! {
+    user_event_excuses (id) {
+          id -> Integer,
+          user_id -> Integer,
```
This is inconsistent with the SQL migration. In the migration, it's named `static12_id` but in the Rust definition it's being called `user_id`. I think it should be `legacy_player_id` in-line with our Titan user parlance.

MichaelJamesParsons replied:
The `static12_id` field was meant to be replaced with `user_id` in an earlier refactor. Do we want to keep the legacy id with each excuse?

ELD replied:
I'm alright with replacing the `static12_id` with `user_id` and using the Titan `users` table's `id` field to join on it. You just need to make this consistent with the migration.

MichaelJamesParsons commented:
@ELD Ready for another review.

MichaelJamesParsons commented:
> Only people who can add file entries can add excuses.
That is an oversight. There should be a separate option for excuses.

> However, I can't submit my own excuse if I don't have that ACL permission.
This will be addressed in another PR, then rebased into this branch. This PR is already far too large.

ELD commented:
> That is an oversight. There should be a separate option for excuses.

Okay, sounds good. I'll wait to merge this until you land that fix.

> This will be addressed in another PR, then rebased into this branch. This PR is already far too large.

I disagree. I don't think this PR is too large as-is or with the two pieces of acceptance criteria that were missed.

I'll compromise this once on the process and merge this PR when you've fixed the ACL oversight. I'm doing this to ensure that we ship excuses at the end of this sprint. This won't close #72 or #73 as those are still unfinished.

In the future, if this happens again, I'm going to hold firm. If you think an issue is too large for a single PR, we need to break that issue down. Leaving out functionality specified in the acceptance criteria for an issue is akin to making a unilateral product decision without discussing it first with the product owner.

MichaelJamesParsons commented:
@ELD ready for review.

## Changes

>  >That is an oversight. There should be a separate option for excuses.
>
> Okay, sounds good. I'll wait to merge this until you land that fix.

Added `canCreateEventExcuse` ACL permission.

> > I left out success/error indicators for now to keep the PR small(er). I'll add a snackbar to both the excuses and file entry UI's soon.
> 
> This isn't what was asked for in this feature and should have been included. If you still want it to be part of a second PR for the sake of size, that's okay. I'm going to hold off on merging this until that other PR is ready.

The react material team recently [released a library](https://material-ui.com/demos/snackbars/#notistack) for handling snackbar queuing. Since the necessary boilerplate code is covered by that library, I added the success/error states to this PR.

> I can't submit my own excuse if I don't have that ACL permission.

The `WithAcl` component now supports this through the `authUserOptions` prop.

## Feedback

> > This will be addressed in another PR, then rebased into this branch. This PR is already far too large.
>
> I disagree. I don't think this PR is too large as-is or with the two pieces of acceptance criteria that were missed.

Large PR's can easily hide bugs, flaws in design, and inconsistencies in code quality. I advise we break future features into individual components instead of full views.

- A PR for listing excuses.
- Another for creating excuses.

In the future, PR feedback should be presented in its simplest form and limited to requests for specific changes, per Github's community [feedback guidlines](https://github.blog/2015-01-21-how-to-write-the-perfect-pull-request/#offering-feedback). Other concerns should be addressed elsewhere.
